File: android/app/src/main/kotlin/meditofoundation/medito/SessionBellSchedule.kt
package meditofoundation.medito

/** Cue state uses media time. A seek consumes skipped cues, zero restarts. */
internal class SessionBellSchedule {
    private var started = false
    private var middle = false
    private var end = false

    private fun endPosition(duration: Long, speed: Double): Long =
        maxOf(duration / 2, duration - (6000 * speed).toLong())

    fun reset() { started = false; middle = false; end = false }

    fun seek(position: Long, duration: Long, speed: Double = 1.0) {
        if (position == 0L) reset() else {
            started = true
            middle = duration > 0 && position >= duration / 2
            end = duration > 0 && position >= endPosition(duration, speed)
        }
    }

    fun update(position: Long, duration: Long, speed: Double = 1.0): Boolean {
        if (duration <= 0 || position >= duration) return false
        if (!started) {
            started = true
            middle = position >= duration / 2
            end = position >= endPosition(duration, speed)
            return position < 2000
        }
        if (!end && position >= endPosition(duration, speed)) {
            end = true
            middle = true
            return true
        }
        if (!middle && position >= duration / 2 && position < duration) {
            middle = true
            return true
        }
        return false
    }
}

File: lib/services/audio/session_bell_schedule.dart
/// One-shot cues follow media time, never wall-clock time. Seeking past a cue
/// consumes it; seeking to the beginning starts a fresh session.
class SessionBellSchedule {
  bool _started = false;
  bool _middle = false;
  bool _end = false;

  // The asset lasts six seconds, including its silent tail. Convert that
  // wall-clock duration to media time so faster playback still leaves room.
  Duration _endPosition(Duration duration, double speed) {
    final lead = Duration(microseconds: (6000000 * speed).round());
    final position = duration - lead;
    return position > duration ~/ 2 ? position : duration ~/ 2;
  }

  void reset() {
    _started = false;
    _middle = false;
    _end = false;
  }

  void seek(Duration position, Duration duration, {double speed = 1}) {
    if (position == Duration.zero) {
      reset();
    } else {
      _started = true;
      _middle = duration > Duration.zero && position >= duration ~/ 2;
      _end =
          duration > Duration.zero && position >= _endPosition(duration, speed);
    }
  }

  bool update(Duration position, Duration duration, {double speed = 1}) {
    if (duration <= Duration.zero || position >= duration) return false;
    if (!_started) {
      _started = true;
      _middle = position >= duration ~/ 2;
      _end = position >= _endPosition(duration, speed);
      return position < const Duration(seconds: 2);
    }
    if (!_end && position >= _endPosition(duration, speed)) {
      _end = true;
      _middle = true;
      return true;
    }
    if (!_middle && position >= duration ~/ 2 && position < duration) {
      _middle = true;
      return true;
    }
    return false;
  }
}
